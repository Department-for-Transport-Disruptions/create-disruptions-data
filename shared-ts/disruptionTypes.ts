import { z } from "zod";
import {
    consequenceOperatorsSchema,
    consequenceSchema,
    disruptionInfoSchema,
    disruptionSchema,
    networkConsequenceSchema,
    operatorConsequenceSchema,
    serviceSchema,
    servicesConsequenceSchema,
    stopSchema,
    stopsConsequenceSchema,
    validitySchema,
} from "./disruptionTypes.zod";

export type Disruption = z.infer<typeof disruptionSchema>;
export type DisruptionInfo = z.infer<typeof disruptionInfoSchema>;

export type Validity = z.infer<typeof validitySchema>;

export type Consequence = z.infer<typeof consequenceSchema>;
export type ConsequenceOperators = z.infer<typeof consequenceOperatorsSchema>;
export type Stop = z.infer<typeof stopSchema>;
export type Service = z.infer<typeof serviceSchema>;

export type NetworkConsequence = z.infer<typeof networkConsequenceSchema>;
export type OperatorConsequence = z.infer<typeof operatorConsequenceSchema>;
export type StopsConsequence = z.infer<typeof stopsConsequenceSchema>;
export type ServicesConsequence = z.infer<typeof servicesConsequenceSchema>;

export const routesSchema = z.object({
    inbound: z.array(stopSchema.partial()),
    outbound: z.array(stopSchema.partial()),
});

export type Routes = z.infer<typeof routesSchema>;
