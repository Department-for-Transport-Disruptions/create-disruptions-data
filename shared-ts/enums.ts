export enum SourceType {
    directReport = "directReport",
    email = "email",
    phone = "phone",
    fax = "fax",
    post = "post",
    feed = "feed",
    radio = "radio",
    tv = "tv",
    web = "web",
    pager = "pager",
    text = "text",
    other = "other",
}

export enum Progress {
    draft = "draft",
    open = "open",
    published = "published",
    closing = "closing",
    closed = "closed",
    pendingApproval = "pendingApproval",
    editPendingApproval = "editPendingApproval",
    draftPendingApproval = "draftPendingApproval",
    rejected = "rejected",
}

export enum MiscellaneousReason {
    accident = "accident",
    securityAlert = "securityAlert",
    congestion = "congestion",
    roadClosed = "roadClosed",
    incident = "incident",
    routeDiversion = "routeDiversion",
    unknown = "unknown",
    vandalism = "vandalism",
    overcrowded = "overcrowded",
    operatorCeasedTrading = "operatorCeasedTrading",
    vegetation = "vegetation",
    roadworks = "roadworks",
    specialEvent = "specialEvent",
    insufficientDemand = "insufficientDemand",
}

export enum PersonnelReason {
    unknown = "unknown",
    staffSickness = "staffSickness",
    staffInjury = "staffInjury",
    contractorStaffInjury = "contractorStaffInjury",
    staffAbsence = "staffAbsence",
    staffInWrongPlace = "staffInWrongPlace",
    staffShortage = "staffShortage",
    industrialAction = "industrialAction",
    unofficialIndustrialAction = "unofficialIndustrialAction",
    workToRule = "workToRule",
    undefinedPersonnelProblem = "undefinedPersonnelProblem",
}

export enum EquipmentReason {
    unknown = "unknown",
    pointsFailure = "pointsFailure",
    signalProblem = "signalProblem",
    trainWarningSystemProblem = "trainWarningSystemProblem",
    trackCircuitProblem = "trackCircuitProblem",
    signalFailure = "signalFailure",
    derailment = "derailment",
    engineFailure = "engineFailure",
    tractionFailure = "tractionFailure",
    breakDown = "breakDown",
    technicalProblem = "technicalProblem",
    brokenRail = "brokenRail",
    poorRailConditions = "poorRailConditions",
    wheelImpactLoad = "wheelImpactLoad",
    lackOfOperationalStock = "lackOfOperationalStock",
    defectiveFireAlarmEquipment = "defectiveFireAlarmEquipment",
    defectivePlatformEdgeDoors = "defectivePlatformEdgeDoors",
    defectiveCctv = "defectiveCctv",
    defectivePublicAnnouncementSystem = "defectivePublicAnnouncementSystem",
    ticketingSystemNotAvailable = "ticketingSystemNotAvailable",
    repairWork = "repairWork",
    constructionWork = "constructionWork",
    maintenanceWork = "maintenanceWork",
    emergencyEngineeringWork = "emergencyEngineeringWork",
    lateFinishToEngineeringWork = "lateFinishToEngineeringWork",
    powerProblem = "powerProblem",
    fuelProblem = "fuelProblem",
    swingBridgeFailure = "swingBridgeFailure",
    escalatorFailure = "escalatorFailure",
    liftFailure = "liftFailure",
    gangwayProblem = "gangwayProblem",
    closedForMaintenance = "closedForMaintenance",
    fuelShortage = "fuelShortage",
    deicingWork = "deicingWork",
    wheelProblem = "wheelProblem",
    luggageCarouselProblem = "luggageCarouselProblem",
    undefinedEquipmentProblem = "undefinedEquipmentProblem",
}

export enum EnvironmentReason {
    unknown = "unknown",
    fog = "fog",
    roughSea = "roughSea",
    heavySnowFall = "heavySnowFall",
    driftingSnow = "driftingSnow",
    blizzardConditions = "blizzardConditions",
    heavyRain = "heavyRain",
    strongWinds = "strongWinds",
    stormConditions = "stormConditions",
    stormDamage = "stormDamage",
    tidalRestrictions = "tidalRestrictions",
    highTide = "highTide",
    lowTide = "lowTide",
    ice = "ice",
    frozen = "frozen",
    hail = "hail",
    sleet = "sleet",
    highTemperatures = "highTemperatures",
    flooding = "flooding",
    waterlogged = "waterlogged",
    lowWaterLevel = "lowWaterLevel",
    highWaterLevel = "highWaterLevel",
    fallenLeaves = "fallenLeaves",
    fallenTree = "fallenTree",
    landslide = "landslide",
    undefinedEnvironmentalProblem = "undefinedEnvironmentalProblem",
    lightningStrike = "lightningStrike",
    sewerOverflow = "sewerOverflow",
    grassFire = "grassFire",
}

export enum DayType {
    monday = "monday",
    tuesday = "tuesday",
    wednesday = "wednesday",
    thursday = "thursday",
    friday = "friday",
    saturday = "saturday",
    sunday = "sunday",
}

export enum VehicleMode {
    bus = "bus",
    tram = "tram",
    ferryService = "ferryService",
    rail = "rail",
    underground = "underground",
}

export enum Severity {
    unknown = "unknown",
    normal = "normal",
    verySlight = "verySlight",
    slight = "slight",
    severe = "severe",
    verySevere = "verySevere",
}

export enum UserGroups {
    systemAdmins = "system-admins",
    orgAdmins = "org-admins",
    orgPublishers = "org-publishers",
    orgStaff = "org-staff",
    operators = "operators",
}

export enum PublishStatus {
    draft = "DRAFT",
    published = "PUBLISHED",
    editing = "EDITING",
    pendingApproval = "PENDING_APPROVAL",
    rejected = "REJECTED",
    editPendingApproval = "EDIT_PENDING_APPROVAL",
    pendingAndEditing = "PENDING_EDITING",
}

export enum SocialMediaPostStatus {
    pending = "Pending",
    successful = "Successful",
    rejected = "Rejected",
}

export enum Datasource {
    tnds = "tnds",
    bods = "bods",
}

export enum Modes {
    bus = "bus",
    coach = "coach",
    tram = "tram",
    ferry = "ferry",
    rail = "rail",
    underground = "underground",
    metro = "metro",
}
