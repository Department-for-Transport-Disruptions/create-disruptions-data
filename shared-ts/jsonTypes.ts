
export enum SourceType {
    directReport,
    email,
    phone,
    ax,
    post,
    feed,
    radio,
    tv,
    web,
    pager,
    text,
    other,
}

export enum Progress {
    draft, 
    pendingApproval, 
    approvedDraft, 
    open, 
    closing, 
    closed, 
    rejected
}

export enum MiscellaneousReason {
    accident,
    securityAlert,
    congestion,
    roadWorks,
    roadClosed,
    bombExplosion,
    incident,
    routeDiversion,
    staffAbsence,
    signalProblem,
    maintenanceWork,
    constructionWork,
    fog,
    ice,
    heavyRain,
    waterLogged,
    unknown,
    nearMiss,
    safetyViolation,
    signalPassedAtDanger,
    stationOverrun,
    trainDoor,
    emergencyServicesCall,
    policeRequest,
    fireBrigadeSafetyChecks,
    unattendedBag,
    telephonedThreat,
    suspectVehicle,
    civilEmergency,
    airRaid,
    sabotage,
    bombAlert,
    attach,
    evacuation,
    terroristIncident,
    gunfireOnRoadway,
    explosion,
    explosionHazard,
    securityIncident,
    fire,
    linesideFire,
    vandalism,
    passengerAction,
    staffAssault,
    railwayCrime,
    fatality,
    personUnderTrain,
    personHitByTrain,
    personIllOnVehicle,
    emergencyServices,
    collision,
    overcrowded,
    insufficientDemand,
    lightingFailure,
    leaderBoardFailure,
    serviceIndicatorFailure,
    serviceFailure,
    operatorCeasedTrading,
    operatorSuspended,
    routeBlockage,
    personOnTheLine,
    vehicleOnTheLine,
    levelCrossingIncident,
    objectOnTheLine,
    fallenTreeOnTheLine,
    vegetation,
    trainStruckAnimal,
    trainStruckObject,
    animalOnTheLine,
    roadworks,
    specialEvent,
    march,
    procession,
    demonstration,
    publicDisturbance,
    filterBlockade,
    sightseersObstructingAccess,
    bridgeStrike,
    overheadObstruction,
    undefinedProblem,
    problemsAtBorderPost,
    problemsAtCustomsPost,
    problemsOnLocalRoad
}

export enum PersonnelReason {
    unknown,
    staffSickness,
    staffInjury,
    contractorStaffInjury,
    staffAbsence,
    staffInWrongPlace,
    staffShortage,
    industrialAction,
    unofficialIndustrialAction,
    workToRule,
    undefinedPersonnelProblem
}

export enum EquipmentReason {
    unknown,
    pointsProblem,
    pointsFailure,
    signalProblem,
    trainWarningSystemProblem,
    trackCircuitProblem,
    signalFailure,
    derailment,
    engineFailure,
    tractionFailure,
    breakDown,
    technicalProblem,
    brokenRail,
    poorRailConditions,
    wheelImpactLoad,
    lackOfOperationalStock,
    defectiveFireAlarmEquipment,
    defectivePlatformEdgeDoors,
    defectiveCctv,
    defectivePublicAnnouncementSystem,
    ticketingSystemNotAvailable,
    repairWork,
    constructionWork,
    maintenanceWork,
    emergencyEngineeringWork,
    lateFinishToEngineeringWork,
    powerProblem,
    fuelProblem,
    swingBridgeFailure,
    escalatorFailure,
    liftFailure,
    gangwayProblem,
    closedForMaintenance,
    fuelShortage,
    deicingWork,
    wheelProblem,
    luggageCarouselProblem,
    undefinedEquipmentProblem
}

export enum EnvironmentReason {
    unknown,
    fog,
    roughSea,
    heavySnowFall,
    driftingSnow,
    blizzardConditions,
    heavyRain,
    strongWinds,
    stormConditions,
    stormDamage,
    tidalRestrictions,
    highTide,
    lowTide,
    ice,
    frozen,
    hail,
    sleet,
    highTemperatures,
    flooding,
    waterlogged,
    lowWaterLevel,
    highWaterLevel,
    fallenLeaves,
    fallenTree,
    landslide,
    undefinedEnvironmentalProblem,
    lightningStrike,
    sewerOverflow,
    grassFire
}

export enum DayType {
    monday,
    tuesday,
    wednesday,
    thursday,
    friday,
    saturday,
    sunday
}

export type Source = {
    SourceType: SourceType;
    TimeOfCommunication: string;
};

export type Period = {
    StartTime: string;
    EndTime?: string;
};

export type InfoLink = {
    Uri: string;
}


export type Reference = {
    RelatedToRef: string[];
};

export type Repetitions = {
    DayType: DayType[];
}

export type InfoLinks = {
    InfoLink: InfoLink[];
  }

export type BasePtSituationElement = {
    CreationTime: string;
    ParticipantRef: string;
    SituationNumber: string;
    Version?: Number;
    References?: Reference;
    Source: Source;
    Progress: Progress;
    ValidityPeriod: Period[];
    Repetitions?: Repetitions;
    PublicationWindow: Period;
    Planned: boolean;
    Summary: string;
    Description: string;
    InfoLinks?: InfoLinks;
};

export type MiscReasonPtSituationElement = BasePtSituationElement & {
  MiscellaneousReason: MiscellaneousReason;
};

export type PersonnelReasonPtSituationElement = BasePtSituationElement & {
  PersonnelReason: PersonnelReason;
};

export type EquipmentReasonPtSituationElement = BasePtSituationElement & {
  EquipmentReason: EquipmentReason;
};

export type EnvironmentReasonPtSituationElement = BasePtSituationElement & {
  EnvironmentReason: EnvironmentReason;
};

export type PtSituationElement = MiscReasonPtSituationElement | PersonnelReasonPtSituationElement | EquipmentReasonPtSituationElement | EnvironmentReasonPtSituationElement;

export type SituationExchangeDelivery = {
    ResponseTimestamp: string;
    Status?: boolean;
    ShortestPossibleCycle?: string;
    Situations: PtSituationElement[];
}

export type ServiceDelivery = {
    ResponseTimestamp: string;
    ProducerRef: string;
    ResponseMessageIdentifier: string;
    SituationExchangeDelivery: SituationExchangeDelivery;
}