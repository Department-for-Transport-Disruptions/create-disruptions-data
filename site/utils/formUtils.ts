import dayjs from "dayjs";
import { Dispatch, SetStateAction } from "react";
import { z } from "zod";
import { getDate } from "./dates";
import { ErrorInfo, PageState } from "../interfaces";
import { ConsequenceOperators, ServiceApiResponse, Stop } from "../schemas/consequence.schema";
import { sortServices } from ".";

export const handleBlur = <T>(
    input: string,
    inputName: Extract<keyof T, string>,
    stateUpdater: (change: string, field: keyof T) => void,
    setErrors: Dispatch<SetStateAction<ErrorInfo[]>>,
    schema?: z.ZodTypeAny,
    disabled?: boolean,
) => {
    stateUpdater(input, inputName);

    if (schema && !disabled) {
        const parsed = schema.safeParse(input);

        if (parsed.success === false) {
            setErrors([
                {
                    id: inputName,
                    errorMessage: parsed.error.errors[0].message,
                },
            ]);
        } else {
            setErrors([]);
        }
    }
};

export const getStateUpdater =
    <T>(setter: (value: SetStateAction<PageState<Partial<T>>>) => void, state: PageState<Partial<T>>) =>
    (change: string | string[], field: keyof T) => {
        setter({
            ...state,
            inputs: {
                ...state.inputs,
                [field]: change,
            },
        });
    };

export const operatorStateUpdater =
    <T>(setter: (value: SetStateAction<PageState<Partial<T>>>) => void, state: PageState<Partial<T>>) =>
    (change: ConsequenceOperators[], field: keyof T) => {
        setter({
            ...state,
            inputs: {
                ...state.inputs,
                [field]: change,
            },
        });
    };

export const getStopLabel = (stop: Stop) => {
    if (stop.commonName && stop.indicator && stop.atcoCode) {
        return `${stop.commonName} (${stop.indicator}) (${stop.atcoCode})`;
    } else if (stop.commonName && stop.atcoCode) {
        return `${stop.commonName} (${stop.atcoCode})`;
    } else {
        return "";
    }
};

export const getStopValue = (stop: Stop) => stop.atcoCode.toString();

export const sortStops = (stops: Stop[]) => {
    return stops.sort((a, b) => {
        if (a.commonName && a.indicator && a.atcoCode && b.indicator) {
            return (
                a.commonName.localeCompare(b.commonName) ||
                a.indicator.localeCompare(b.indicator) ||
                a.atcoCode.localeCompare(b.atcoCode)
            );
        } else {
            return a.commonName.localeCompare(b.commonName) || a.atcoCode.localeCompare(b.atcoCode);
        }
    });
};

export const getDataInPages = <T>(pageNumber: number, data: T[]): T[] => {
    const startPoint = (pageNumber - 1) * 10;
    const endPoint = pageNumber * 10;
    return data.slice(startPoint, endPoint);
};

export const getStopType = (stopType: string | undefined) => {
    if (stopType === "BCT") {
        return "Bus stop";
    } else if (stopType === "MET" || stopType === "PLT") {
        return "Tram stop";
    } else if (stopType === "FER" || stopType === "FBT") {
        return "Ferry terminal";
    } else {
        return "Stop";
    }
};

export const filterServices = (servicesData?: ServiceApiResponse[]) => {
    if (!servicesData?.length) {
        return [];
    }

    const services = sortServices(servicesData);
    const filterKey = services[0].dataSource === "tnds" ? "serviceCode" : "lineId";

    return removeDuplicateServicesByKey(services, filterKey);
};

export const removeDuplicateServicesByKey = (services: ServiceApiResponse[], filterKey: "serviceCode" | "lineId") => {
    const setOfServiceIds = new Set();
    const filteredServices: ServiceApiResponse[] = [];
    const currentDate = getDate();

    services.forEach((currentService) => {
        const endDate = currentService.endDate === null ? currentDate.add(1, "day") : dayjs(currentService.endDate);
        if (!setOfServiceIds.has(currentService[filterKey])) {
            setOfServiceIds.add(currentService[filterKey]);
            filteredServices.push(currentService);
        } else {
            if (currentDate.isBetween(dayjs(currentService.startDate), endDate, "day", "[]")) {
                const serviceToReplace = filteredServices.findIndex(
                    (serviceToReplace) => serviceToReplace[filterKey] === currentService[filterKey],
                );
                filteredServices[serviceToReplace] = currentService;
            }
        }
    });
    return filteredServices;
};
