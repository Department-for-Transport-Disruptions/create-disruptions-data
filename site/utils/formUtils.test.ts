import { Stop } from "@create-disruptions-data/shared-ts/disruptionTypes";
import MockDate from "mockdate";
import { afterEach, describe, expect, it } from "vitest";
import {
    mockBodsServicesNoDuplicates,
    mockBodsServicesWithDuplicates,
    mockTndsServicesNoDuplicates,
    mockTndsServicesWithDuplicates,
} from "../testData/mockData";
import { filterServices, getStopType, removeDuplicateServicesByKey, sortAndFilterStops } from "./formUtils";

describe("formUtils", () => {
    afterEach(() => {
        MockDate.reset();
    });

    describe("filterServices", () => {
        it.each([[mockBodsServicesNoDuplicates], [mockBodsServicesWithDuplicates]])(
            "should return a sorted array of unique services if the data source is bods",
            (services) => {
                MockDate.set("2023-08-08");
                expect(filterServices(services)).toMatchSnapshot();
            },
        );
        it.each([[mockTndsServicesNoDuplicates], [mockTndsServicesWithDuplicates]])(
            "should return a sorted array of unique services if the data source is tnds",
            (services) => {
                MockDate.set("2023-08-08");
                expect(filterServices(services)).toMatchSnapshot();
            },
        );
        it("should return an empty array if no data is passed", () => {
            expect(filterServices([])).toEqual([]);
        });
    });

    describe("removeDuplicateServices", () => {
        it.each([[mockBodsServicesNoDuplicates], [mockBodsServicesWithDuplicates]])(
            "should return an array of unique services filtered by lineId if the data source is bods",
            (services) => {
                MockDate.set("2023-08-08");
                expect(removeDuplicateServicesByKey(services, "lineId")).toMatchSnapshot();
            },
        );
        it.each([[mockTndsServicesNoDuplicates], [mockTndsServicesWithDuplicates]])(
            "should return an array of unique services filtered by ServiceCode if the data source is tnds",
            (services) => {
                MockDate.set("2023-08-08");
                expect(removeDuplicateServicesByKey(services, "serviceCode")).toMatchSnapshot();
            },
        );
    });

    describe("sortFilterStops", () => {
        const unsortedDuplicatedStops: Stop[] = [
            {
                atcoCode: "370020440",
                commonName: "Prospect Road/St Quentin Drive",
                indicator: "adj",
                bearing: "E",
                longitude: -1.512933989,
                latitude: 53.317638172,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "inbound",
                sequenceNumber: "2",
            },
            {
                atcoCode: "370025332",
                commonName: "Spa Lane/Tannery Street",
                indicator: "adj",
                bearing: "S",
                longitude: -1.368099159,
                latitude: 53.357876345,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "outbound",
                sequenceNumber: "2",
            },
            {
                atcoCode: "370020349",
                commonName: "Cross Street/Tannery Street",
                indicator: "at",
                bearing: "S",
                longitude: -1.370176766,
                latitude: 53.358615412,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "outbound",
                sequenceNumber: "1",
            },
            {
                atcoCode: "370025442",
                commonName: "Prospect Road/Everard Avenue",
                indicator: "adj",
                bearing: "SE",
                longitude: -1.516408243,
                latitude: 53.318371398,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "inbound",
                sequenceNumber: "1",
            },
            {
                atcoCode: "370025334",
                commonName: "Spa Lane/Spa Lane Croft",
                indicator: "at",
                bearing: "SW",
                longitude: -1.368838473,
                latitude: 53.35563306,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "outbound",
                sequenceNumber: "3",
            },
            {
                atcoCode: "370025334",
                commonName: "Spa Lane/Spa Lane Croft",
                indicator: "at",
                bearing: "SW",
                longitude: -1.368838473,
                latitude: 53.35563306,
                stopType: "BCT",
                busStopType: "MKD",
                direction: "outbound",
                sequenceNumber: "3",
            },
        ];

        it("should return a list of sorted stops", () => {
            expect(sortAndFilterStops(unsortedDuplicatedStops)).toMatchSnapshot();
        });
        it("should an empty array when there are no stops passed in", () => {
            expect(sortAndFilterStops([])).toEqual([]);
        });
    });

    describe("getStopType", () => {
        it.each([
            ["BCT", "Bus stop"],
            ["BCS", "Bus stop"],
            ["BCQ", "Bus stop"],
            ["MET", "Tram stop"],
            ["PLT", "Tram stop"],
            ["FER", "Ferry terminal"],
            ["FBT", "Ferry terminal"],
            [undefined, "Stop"],
        ])("should return the correct name for a given stop type", (stopType, stopName) => {
            expect(getStopType(stopType)).toEqual(stopName);
        });
    });
});
