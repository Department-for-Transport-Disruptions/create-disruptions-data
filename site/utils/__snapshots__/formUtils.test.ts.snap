// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formUtils > filterServices > should return a sorted array of unique services if the data source is bods 1`] = `
[
  {
    "dataSource": "bods",
    "destination": "Warrington",
    "endDate": "2023-09-23",
    "id": 4524,
    "lineId": "LineId2",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-09-01",
  },
  {
    "dataSource": "bods",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "LineId1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "LineId3",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > filterServices > should return a sorted array of unique services if the data source is bods 2`] = `
[
  {
    "dataSource": "bods",
    "destination": "Warrington",
    "endDate": "2023-08-10",
    "id": 4524,
    "lineId": "LineId2",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Oldham circular",
    "endDate": "2023-08-08",
    "id": 4513,
    "lineId": "LineId1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "LineId3",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > filterServices > should return a sorted array of unique services if the data source is tnds 1`] = `
[
  {
    "dataSource": "tnds",
    "destination": "Warrington",
    "endDate": "2023-09-23",
    "id": 4524,
    "lineId": "SL1",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-09-01",
  },
  {
    "dataSource": "tnds",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "SL1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "SL1",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > filterServices > should return a sorted array of unique services if the data source is tnds 2`] = `
[
  {
    "dataSource": "tnds",
    "destination": "Warrington",
    "endDate": "2023-08-10",
    "id": 4524,
    "lineId": "SL1",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "SL1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "SL1",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > removeDuplicateServices > should return an array of unique services filtered by ServiceCode if the data source is tnds 1`] = `
[
  {
    "dataSource": "tnds",
    "destination": "Warrington",
    "endDate": "2023-09-23",
    "id": 4524,
    "lineId": "SL1",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-09-01",
  },
  {
    "dataSource": "tnds",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "SL1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "SL1",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > removeDuplicateServices > should return an array of unique services filtered by ServiceCode if the data source is tnds 2`] = `
[
  {
    "dataSource": "tnds",
    "destination": "Warrington",
    "endDate": "2023-08-10",
    "id": 4524,
    "lineId": "SL1",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "SL1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "tnds",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "SL1",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > removeDuplicateServices > should return an array of unique services filtered by lineId if the data source is bods 1`] = `
[
  {
    "dataSource": "bods",
    "destination": "Warrington",
    "endDate": "2023-09-23",
    "id": 4524,
    "lineId": "LineId2",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-09-01",
  },
  {
    "dataSource": "bods",
    "destination": "Oldham circular",
    "endDate": "2023-08-10",
    "id": 4513,
    "lineId": "LineId1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "LineId3",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > removeDuplicateServices > should return an array of unique services filtered by lineId if the data source is bods 2`] = `
[
  {
    "dataSource": "bods",
    "destination": "Warrington",
    "endDate": "2023-08-10",
    "id": 4524,
    "lineId": "LineId2",
    "lineName": "100",
    "nocCode": "GONW",
    "operatorShortName": "Go North West",
    "origin": "Manchester",
    "serviceCode": "NW_04_GONW_100_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Oldham circular",
    "endDate": "2023-08-08",
    "id": 4513,
    "lineId": "LineId1",
    "lineName": "149",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "NMGH",
    "serviceCode": "NW_04_SCMN_149_1",
    "startDate": "2023-07-23",
  },
  {
    "dataSource": "bods",
    "destination": "Manchester",
    "endDate": null,
    "id": 4547,
    "lineId": "LineId3",
    "lineName": "253",
    "nocCode": "SCMN",
    "operatorShortName": "Stagecoach",
    "origin": "Partington",
    "serviceCode": "NW_04_SCMN_253_1",
    "startDate": "2023-08-15",
  },
]
`;

exports[`formUtils > sortFilterStops > should return a list of sorted stops 1`] = `
[
  {
    "atcoCode": "370020349",
    "bearing": "S",
    "busStopType": "MKD",
    "commonName": "Cross Street/Tannery Street",
    "direction": "outbound",
    "indicator": "at",
    "latitude": 53.358615412,
    "longitude": -1.370176766,
    "sequenceNumber": "1",
    "stopType": "BCT",
  },
  {
    "atcoCode": "370025442",
    "bearing": "SE",
    "busStopType": "MKD",
    "commonName": "Prospect Road/Everard Avenue",
    "direction": "inbound",
    "indicator": "adj",
    "latitude": 53.318371398,
    "longitude": -1.516408243,
    "sequenceNumber": "1",
    "stopType": "BCT",
  },
  {
    "atcoCode": "370020440",
    "bearing": "E",
    "busStopType": "MKD",
    "commonName": "Prospect Road/St Quentin Drive",
    "direction": "inbound",
    "indicator": "adj",
    "latitude": 53.317638172,
    "longitude": -1.512933989,
    "sequenceNumber": "2",
    "stopType": "BCT",
  },
  {
    "atcoCode": "370025334",
    "bearing": "SW",
    "busStopType": "MKD",
    "commonName": "Spa Lane/Spa Lane Croft",
    "direction": "outbound",
    "indicator": "at",
    "latitude": 53.35563306,
    "longitude": -1.368838473,
    "sequenceNumber": "3",
    "stopType": "BCT",
  },
  {
    "atcoCode": "370025332",
    "bearing": "S",
    "busStopType": "MKD",
    "commonName": "Spa Lane/Tannery Street",
    "direction": "outbound",
    "indicator": "adj",
    "latitude": 53.357876345,
    "longitude": -1.368099159,
    "sequenceNumber": "2",
    "stopType": "BCT",
  },
]
`;
